Documentation: Add documentation for hive (#23746)
cilium-team/cilium: Documentation/contributing/development/hive.rst

.. only:: not (epub or latex or html)

    WARNING: You are looking at unreleased Cilium documentation.
    Please use the official rendered version released here:
    https://docs.cilium.io

.. _hive:

Guide to the Hive
=================

Introduction
~~~~~~~~~~~~

> Cilium is using dependency injection (via ``pkg/hive``) to wire up the

Reviewer: I would really like to have a one-sentence summary on dependency injection with reflection, and/or a link to learn more on that topic. I don't think all readers will be familiar with the concept.

Author: Added both.

initialization, starting and stopping of its components. Dependency injection
helps with modularity by automating the process of initializing, introducing,
starting and stopping of components. Dependency injection also reduces
the inclination to use global variables which are often source of bugs
(due to initialization order) and difficult to deal with in tests (as the
state needs to be restored). Dependency injection lifts code into "data",
enabling visualization of inter-component dependencies and thus making the
code easier to navigate.

Dependency injection and the machinery described here are only a tool to
help us towards the real goal: a modular software architecture that can be
easily understood, extended, repurposed, tested and refactored by a large
group of developers with minimal overlap between modules. To achieve this we
also need to have modularity in mind when designing the architecture and APIs.

Hive and Cells
~~~~~~~~~~~~~~

Cilium applications are composed using runtime dependency injection from a set
of modular components called cells that compose together to form a hive (as in

> bee hive). A hive can be then supplied with configuration and executed. To provide

Reviewer 2: ```suggestion
bee hive). A hive can then be supplied with configuration and executed. To provide
```

a feel for what this is about, here is how a simple modular HTTP server application 
would leverage hive:

.. code-block:: go

    package server

>     // Cell implements a generic HTTP server. Provides the 'Server' API

Reviewer 2: At this point, saying just `Cell implements ...` it sounds like you're talking about Cells in general. Maybe

```suggestion
    // This cell implements a generic HTTP server. Provides the 'Server' API
```

I briefly thought calling it `ServerCell` would be better, but it stutters (`server.ServerCell`). So maybe differentiate Identifiers from concepts using quoting, maybe \`Cell\` implements...?

Author: changed it to `The server cell implements...` and kept it as `server.Cell`

    // for registering request handlers.
    //
    // Module() creates a named collection of cells.
    var Cell = cell.Module(
       "http-server", // Module identifier (for e.g. logging and tracing)
       "HTTP Server", // Module title (for documentation)

       // Provide "provides" the application with a constructor for one or more

>        // objects. It is lazy and won't be called unless Invoke()'d (directly, or

Reviewer 2: On the fence about this one, but calling out the lazyness at this point already I think is a bit premature. I think the purpose of `Provide` is enough to take in at this stage, and the lazyness can come up in the more detailed intro to it.

In addition, I think here it could just be `Provide the application with a constructor for the server`, ie keep it concrete to the example.

Author: Yeah I think it's fine to drop mentioning that it's lazy. I'm anyway mentioning it later.

       // as a dependency of another constructor).
       cell.Provide(New),

       // Config registers a configuration when provided with the defaults 
       // and an implementation of Flags() for registering the configuration flags.
       cell.Config(defaultServerConfig),
    )

    // Server allows registering request handlers with the HTTP server
    type Server interface {
        ListenAddress() string
        RegisterHandler(path string, fn http.HandlerFunc)
    }

    func New(lc hive.Lifecycle, cfg ServerConfig) Server { 
      // Initialize http.Server, register Start and Stop hooks to Lifecycle 
      // for starting and stopping the server and return an implementation of
      // 'Server' for other cells for registering handlers.
      // ...
    }

    type ServerConfig struct {
        ServerPort uint16
    }

    var defaultServerConfig = ServerConfig{
        ServerPort: 8080,
    }

    func (def ServerConfig) Flags(flags *pflag.FlagSet) {

>         // Register the "server-port" flag. Hive by default drops the "-" and does

Reviewer 2: I think we can keep it concrete here too,
```suggestion
        // Register the "server-port" flag. Hive maps the flag to the
        // ServerPort field in the ServerConfig struct.
```

(Suggestions just for clarity, they don't apply cleanly as is)

        // a case-insensitive comparison with the fields in the configuration struct
        // to fill it from the command-line flags.

>         flags.Int("server-port",  def.ServerPort, "Sets the HTTP server listen port")

Reviewer 3: ```suggestion
        flags.Uint16("server-port", def.ServerPort, "Sets the HTTP server listen port")
```

    }

> 

Reviewer 2: mabe a quick sentence that this is now a separate package and file, something like

Let's now look at a consumer of this cell, the `hello` package.

Author: :+1:

.. code-block:: go

    package hello

    // Cell registers a hello handler to the HTTP server.
    //
    // This cell isn't a Module, but rather just a plain Invoke. An Invoke
    // is a cell that, unlike Provide, is always executed. Invoke functions
    // can depend on values that constructors registered with Provide() can

>     // return. These constructors are then called and their results remembered.

Reviewer 2: mhh maybe this makes it tricky to remove the lazyness intro above. I still think we'd gain clarity but unsure how to then explain this invoke. The Invoke, in general, I think is the most difficult to grasp intuitively based on the name. 🤔

    var Cell = cell.Invoke(registerHelloHandler)

    func helloHandler(w http.ResponseWriter, req *http.Request) {
        w.Write([]byte("hello"))
    }

    func registerHelloHandler(srv server.Server) {
        srv.RegisterHandler("/hello", helloHandler)
    }

>   

Reviewer 2: how about another quick break, maybe

Now, to put it all together, let's combine our cells into hive in the `main` package.

Author: Added couple breaks.

        // exampleHive when executed.
        cmd *cobra.Command = &cobra.Command{
            Use: "example",
            Run: func(cmd *cobra.Command, args []string) {
                // Run() will execute all invoke functions, followed by start hooks
                // and will then wait for interrupt signal before executing stop hooks
                // and returning.
                exampleHive.Run()
            },
        }
    )
       
    func main() {
         // Register all command-line flags from each config cell to the
         // flag-set of our command.
     	 exampleHive.RegisterFlags(cmd.Flags())

         // Add the "hive" sub-command for inspecting the application. 
         cmd.AddCommand(exampleHive.Command()))

         // Execute the root command.
         cmd.Execute()
    }

If you prefer to learn by example you can find a more complete and runnable example
application from ``pkg/hive/example``. Try running it with ``go run .`` and also try
``go run . hive``. And if you're interested in how all this is implemented internally,
see ``pkg/hive/example/mini``, a minimal example of how to do dependency injection with reflection.

The Hive API
~~~~~~~~~~~~

> With the example hopefully having now whetted the appetite, we'll now take a proper look at

Reviewer: Nit: repetition on `now`.

the hive API. 

> `pkg/hive <https://pkg.go.dev/github.com/cilium/cilium@v1.13.0-rc5/pkg/hive>`_ provides the Hive type and 

Reviewer: Should we pin to a particular version here? I understand the risk of getting the link broken, but at the same time, we'll never remember to update the version here and this will get outdated, so I'm torn.

If pinning, please update to 1.13 now that the version is out.

Here and at other locations below.

Author: Now that 1.13 is out I can remove the pinning, so I'll do that.

    cmd.AddCommand(hive.Command())

`pkg/hive/cell <https://pkg.go.dev/github.com/cilium/cilium@v1.13.0-rc5/pkg/hive/cell>`_ defines the Cell interface that 
``hive.New()`` consumes and the following functions for creating cells:

- :ref:`api_module`: A named set of cells.
- :ref:`api_provide`: Provides constructor(s) to the hive.  Lazy and only invoked if referenced by an Invoke function (directly or indirectly via other constructor).
- :ref:`ProvidePrivate <api_module>`: Provides private constructor(s) to a module and its sub-modules.
- :ref:`api_decorate`: Wraps a set of cells with a decorator function to provide these cells with augmented objects.
- :ref:`api_config`: Provides a configuration struct to the hive. Can be registered to a flag set with Hive.RegisterFlags().
- :ref:`api_invoke`: Registers an invoke function to instantiate and initialize objects.

Hive also by default provides the following globally available objects:

- :ref:`api_lifecycle`: Methods for registering Start and Stop functions that are executed when Hive is started and stopped. 
  The hooks are appended to it in dependency order (since the constructors are invoked in dependency order).

> - *hive.Shutdowner*: Allows gracefully shutting down the hive from anywhere in case of a fatal error post-start.

Reviewer 2: The italics look a bit weird because the colon isn't included. Either drop them/use bold instead (typographically not great though). The others are visually nicely distinct because of the link color- is there no links for these?

Reviewer: Don't use bold or italics, but please format as literals to indicate this is code (` ``hive.Shoutdowner`` `). This way you also won't need to declare these as exceptions for the spell checker. I'd use it at most occurrences, save maybe in titles or when you have links on the terms.

- *logrus.FieldLogger*: Interface to the logger. Module() decorates it with ``subsys=<module id>``.

.. _api_provide:

Provide
^^^^^^^

We'll now take a look at each of the different kinds of cells, starting with Provide(),
which registers one or more constructors with the hive:

.. code-block:: go

    // func Provide(ctors any...) Cell

    type A interface {}
    func NewA() A { return A{} }
    
    type B interface {}
    func NewB(A) B { return B{} }

    // simpleCell provides A and B
    var simpleCell cell.Cell = cell.Provide(NewA, NewB) 

If the constructors take many parameters, we'll want to group them into a struct with ``cell.In``,
and conversely if there are many return values, into a struct with ``cell.Out``. This tells
hive to unpack them:

.. code-block:: go

    type params struct {
    	cell.In
    
        A A
        B B
        Lifecycle hive.Lifecycle
    }
    
    type out struct {
        cell.Out
    
        C C
	D D
        E E
    }
    func NewCDE(params params) out { ... }
    
    var Cell = cell.Provide(NewCDE)

Sometimes we want to depend on a group of values sharing the same type, e.g. to collect API handlers or metrics. This can be done with 

> `value groups <https://pkg.go.dev/go.uber.org/dig#hdr-Value_Groups>`_ by combining ``cell.In``

Reviewer 2: both `cell.In` and `cell.Out` support the tag, right?

                 K8sKubeConfigPath: (string) "",
                 K8sClientQPS: (float32) 0,
                 K8sClientBurst: (int) 0,
                 K8sHeartbeatTimeout: (time.Duration) 30s,
                 EnableK8sAPIDiscovery: (bool) false
             }
 
             🚧 client.newClientset (cell.go:109):
                 ⇨ client.Config, hive.Lifecycle, logrus.FieldLogger 
                 ⇦ client.Clientset 
    ...

    Start hooks:

        • gops.registerGopsHooks.func1 (cell.go:44)
        • cmd.newDatapath.func1 (daemon_main.go:1625)
        ...

    Stop hooks:
        ...
   

The hive command prints out the cells, showing what modules, providers,
configurations etc. exist and what they're requiring and providing.
Finally the command prints out all registered start and stop hooks.
Note that these hooks often depend on the configuration (e.g. k8s-client
will not insert a hook unless e.g. --k8s-kubeconfig-path is given). The
hive command takes the same command-line flags as the root command.

The provider dependencies in a hive can also be visualized as a graphviz dot-graph:

.. code-block:: bash

    cilium$ go run ./daemon hive dot-graph | dot -Tx11

> Guidelines

Reviewer 2: These deserve an introduction sentence. Maybe even call out specifically that these can and should be used as Code review comments like https://github.com/golang/go/wiki/CodeReviewComments

Author: :+1:

> Guidelines

Reviewer: Do we have any kind of CI check to enforce these guidelines? To me, some of them look easy to miss, and developers won't always think of checking this list.

Author: There's a test (daemon/cmd/cells_test.go) checks that the hive can be instantiated which potentially catches side-effects in the constructors. For the rest they need to be specifically tested for (e.g. cell's integration test checks with goleak that Stop() cleans up). If I do come up with guidelines and a generic way of testing for it I'll definitely implement it. (Already have some ideas about being able to enforce dependency direction...)

~~~~~~~~~~

* A constructor function should only do validation and allocation. Spawning
  of goroutines or I/O operations must not be performed from constructors,
  but rather via the Start hook. This is required as we want to inspect the
  object graph (e.g. hive.PrintObjects) and side-effectful constructors would
  cause undesired effects.

* Stop functions should make sure to block until all resources
  (goroutines, file handles, …) created by the module have been cleaned
  up (with e.g. sync.WaitGroup). This makes sure that independent
  tests in the same test suite are not affecting each other. Use
  `goleak <https://github.com/uber-go/goleak>`_ to check that goroutines
  are not leaked.

* Each module should come with an integration test that validates that
  it implements its public API correctly. The integration test also serves

>   as an example of how the module’s API is used and it also validates the

Reviewer: `module’s` -> Please avoid mixing straight and curly quotes/apostrophes in the doc. I'd recommend straight everywhere, since it's what people usually expect in text files.

Here and at other locations.

Author: Ah, an annoying artifact from copying the whole text from notion. I'll double check this and also double-quotes as those were funny too.

    c := dig.New(dig.DeferAcyclicVerification())

    // Add in the constructors. Order does not matter.
    c.Provide(NewC)
    c.Provide(NewB)
    c.Provide(NewA)

    // Invoke a function that can depend on any of the values supplied by the
    // registered constructors.
    // Since this depends on “C”, dig will construct first A and B
    // (as C depends on them), and then C.
    c.Invoke(func(c *C) {
        // Do something with C
    })

This is the basis on top of which Hive is built. Hive calls dig’s Provide()
for each of the constructors registered with cell.Provide and then calls
invoke functions to construct the needed objects. The results from the
constructors are memoized, so each constructor is called only once.

``uber/dig`` uses Go’s "reflect" package that provides access to the
type information of the provide and invoke functions. For example, the
`Provide <https://pkg.go.dev/go.uber.org/dig#Container.Provide>`_ method does
something akin to this under the hood:

.. code-block:: go

    // 'constructor' has type "func(...) ..."
    typ := reflect.TypeOf(constructor)
    if typ.Kind() != reflect.Func { /* error */ }

>     in := make([]reflect.Type, typ.NumIn())
>     for i := 0; i < typ.NumIn(); i++ { in = append(in, typ.In(i)) }

Reviewer 4: ```suggestion
    in := make([]reflect.Type, 0, typ.NumIn())
    for i := 0; i < typ.NumIn(); i++ { in = append(in, typ.In(i)) }
```

or

```suggestion
    in := make([]reflect.Type, typ.NumIn())
    for i := 0; i < typ.NumIn(); i++ { in[i] = typ.In(i) }
```

>     out := make([]reflect.Type, typ.NumOut())
>     for i := 0; i < typ.NumOut(); i++ { out = append(out, typ.Out(i)) }

Reviewer 4: ```suggestion
    out := make([]reflect.Type, 0, typ.NumOut())
    for i := 0; i < typ.NumOut(); i++ { out = append(out, typ.Out(i)) }
```

or

```suggestion
    out := make([]reflect.Type, typ.NumOut())
    for i := 0; i < typ.NumOut(); i++ { out[i] = typ.Out(i) }
```

        // exampleHive when executed.
        cmd *cobra.Command = &cobra.Command{
            Use: "example",
            Run: func(cmd *cobra.Command, args []string) {
                // Run() will execute all invoke functions, followed by start hooks
                // and will then wait for interrupt signal before executing stop hooks
                // and returning.
                exampleHive.Run()
            },
        }
    )
       
    func main() {
         // Register all command-line flags from each config cell to the
         // flag-set of our command.
     	 exampleHive.RegisterFlags(cmd.Flags())

         // Add the "hive" sub-command for inspecting the application. 
         cmd.AddCommand(exampleHive.Command()))

         // Execute the root command.
         cmd.Execute()
    }

If you prefer to learn by example you can find a more complete and runnable example
application from ``pkg/hive/example``. Try running it with ``go run .`` and also try
``go run . hive``. And if you're interested in how all this is implemented internally,
see ``pkg/hive/example/mini``, a minimal example of how to do dependency injection with reflection.

The Hive API
~~~~~~~~~~~~

> With the example hopefully having now whetted the appetite, we'll ake a proper look at

Reviewer: “we'll take” (and remove `ake` from spelling exceptions, please)
